sound/soc/intel/boards/ehl_rt5660.c

```diff
@@ -0,0 +1,338 @@
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi2")));
+
+SND_SOC_DAILINK_DEF(idisp3_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("iDisp3 Pin")));
+SND_SOC_DAILINK_DEF(idisp3_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi3")));
+
+SND_SOC_DAILINK_DEF(idisp4_pin,
+	DAILINK_COMP_ARRAY(COMP_CPU("iDisp4 Pin")));
+SND_SOC_DAILINK_DEF(idisp4_codec,
+	DAILINK_COMP_ARRAY(COMP_CODEC("ehdaudio0D2", "intel-hdmi-hifi4")));
+#endif /* IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI) */
+
+static struct snd_soc_dai_link ehl_rt5660_dailink[] = {
+	/* back ends */
+	{
+		.name = "SSP0-Codec",
+		.id = 0,
+		.no_pcm = 1,
+		.dpcm_playback = 1,
+		.dpcm_capture = 1,
+		.ops = &rt5660_ops,
+		.nonatomic = true,
+		SND_SOC_DAILINK_REG(ssp0_pin, rt5660_codec, platform),
+	},
+	{
+		.name = "dmic48k",
+		.id = 1,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic_pin, dmic_codec, platform),
+	},
+	{
+		.name = "dmic16k",
+		.id = 2,
+		.ignore_suspend = 1,
+		.dpcm_capture = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(dmic16k, dmic_codec, platform),
+	},
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
+#define HDMI_LINK_START 3
+#define HDMI_LINE_END 6
+	{
+		.name = "iDisp1",
+		.id = 5,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp1_pin, idisp1_codec, platform),
+	},
+	{
+		.name = "iDisp2",
+		.id = 6,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp2_pin, idisp2_codec, platform),
+	},
+	{
+		.name = "iDisp3",
+		.id = 7,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp3_pin, idisp3_codec, platform),
+	},
+	{
+		.name = "iDisp4",
+		.id = 8,
+		.init = sof_hdmi_init,
+		.dpcm_playback = 1,
+		.no_pcm = 1,
+		SND_SOC_DAILINK_REG(idisp4_pin, idisp4_codec, platform),
+	},
+#endif
+};
+
+/* SoC card */
+static struct snd_soc_card snd_soc_card_ehl_rt5660 = {
+	.name = "ehl-rt5660",
+	.owner = THIS_MODULE,
+	.dai_link = ehl_rt5660_dailink,
+	.num_links = ARRAY_SIZE(ehl_rt5660_dailink),
+	.dapm_widgets = rt5660_widgets,
+	.num_dapm_widgets = ARRAY_SIZE(rt5660_widgets),
+	.dapm_routes = rt5660_map,
+	.num_dapm_routes = ARRAY_SIZE(rt5660_map),
+	.controls = rt5660_controls,
+	.num_controls = ARRAY_SIZE(rt5660_controls),
+	.fully_routed = true,
+	.late_probe = card_late_probe,
+};
+
+#if IS_ENABLED(CONFIG_SND_HDA_CODEC_HDMI)
+/* If hdmi codec is not supported, switch to use dummy codec */
+static void sof_refine_hdmi_links(struct snd_soc_card *card,
+				  struct snd_soc_acpi_mach *mach)
+{
+	struct snd_soc_dai_link *link;
+	int i;
+
+	if (mach->mach_params.common_hdmi_codec_drv &&
+	(mach->mach_params.codec_mask & IDISP_CODEC_MASK))
+		return;
+
+	/* hdmi codec is not supported */
+	for (i = HDMI_LINK_START; i <= HDMI_LINE_END; i++) {
+		link = &card->dai_link[i];
+		link->codecs[0].name = "snd-soc-dummy";
+		link->codecs[0].dai_name = "snd-soc-dummy-dai";
```

Comment:
I see somewhere between a dozen and two dozens of such assignments in ASoC. Each of them allocates a new string with the same contents. Nitpicking (who cares about 200 bytes these days), but we could allocate those two strings centrally and just assign pointers to them to all drivers. Would also make it easier if we need to change them once. Would be a separate PR of course.

Reply:
This should be another topic. My understanding is such strings are allocated statically by the compiler and it will only allocate one time even it is used many times. It should not waste extra memory.

Reply:
@libinyang that would surprise me - you imagine how many strings there are in the kernel, right? You don't expect the compiler to compare all strings across all files? The compiler only compiles one file at a time (including any `#include` additions of course). Those strings are used in different files, so I'd absolutely expect them to create separate copies.

Reply:
@lyakh  This should definitely be a separate PR, but it's a good point. I did a grep and there are 35 copies of these strings in current sof-dev. Worth a general cleanup patch. But not for this PR.

Reply:
> @libinyang that would surprise me - you imagine how many strings there are in the kernel, right? You don't expect the compiler to compare all strings across all files? The compiler only compiles one file at a time (including any `#include` additions of course). Those strings are used in different files, so I'd absolutely expect them to create separate copies.

@lyakh Right. it should be in the same module if they are using the same memory

Reply:
@libinyang not necessarily, that's what `EXPORT_SYMBOL*()` is for

Reply:
so, we are still between two chairs. Either we want to keep HDMI optional, but then you need to use different topologies and have a means to detect which one is relevant, or we make HDMI fixed.
What is the requirement here? I would say for IOT stuff it's rather common to have headless devices so HDMI/DP is rather likely to be used.

Reply:
@plbossart chair is chosen, HDMI is not optional now. @lyakh reducing number of of "snd-soc-dummy" strings is  a separate PR clearly, so I'll mark this discussion resolved, but only in context of this PR